File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        n = len(arr)
        # 左边界 left[i] 为左侧严格小于 arr[i] 的最近元素位置（不存在时为 -1）
        left, st = [-1] * n, []
        for i, x in enumerate(arr):
            while st and arr[st[-1]] >= x:
                st.pop()  # 移除无用数据
            if st: left[i] = st[-1]
            st.append(i)

        # 右边界 right[i] 为右侧小于等于 arr[i] 的最近元素位置（不存在时为 n）
        right, st = [n] * n, []
        for i in range(n - 1, -1, -1):
            while st and arr[st[-1]] > arr[i]:
                st.pop()  # 移除无用数据
            if st: right[i] = st[-1]
            st.append(i)

        ans = 0
        for i, (x, l, r) in enumerate(zip(arr, left, right)):
            ans += x * (i - l) * (r - i)  # 累加贡献
        return ans % (10 ** 9 + 7)

File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.java
class Solution {
    private static final int MOD = 1000000007;
    public int sumSubarrayMins(int[] arr) {
        if (arr.length == 1) return arr[0];
        int left[] = new int[arr.length];
        int right[] = new int[arr.length];
        ArrayDeque<Integer> stack = new ArrayDeque<Integer>(); //the higher the smaller

        for(int i = 0; i < arr.length; i++){
            while(!stack.isEmpty() && arr[stack.peek()] > arr[i]) stack.pop();
            if(stack.isEmpty()) left[i] = -1;
            else left[i] = stack.peek(); //left boundary found
            stack.push(i);
        }

        stack.clear();
        for(int i = arr.length-1; i >= 0; i--){
            while(!stack.isEmpty() && arr[stack.peek()] >= arr[i]) stack.pop();
            if(stack.isEmpty()) right[i] = arr.length;
            else right[i] = stack.peek(); //right boundary found
            stack.push(i);
        }

        long ans = 0;
        for(int i = 0; i < arr.length; i++){
            ans = (ans + (long)(i-left[i])*(right[i]-i)*arr[i]) % MOD;
        }
        return (int)ans;
    }
}